Use ruff instead of flake8, isort, black and pyupgrade

danielhollas opened the pull request:
aiida-core recently switched to ruff, and I've been using it for my app for some time. IMO we should consider slowly switching AiiDAlab repos as well. @unkcpz @yakutovicha LMK what you think. 

I've enabled a couple more rules than there were before, based on my judgment and previous experience. More might be useful, but this is a reasonable start IMO.

I've also removed a couple of hooks that did not seem as useful, now that the codebase and its configuration is quite stable. But I am happy to revert.

danielhollas commented on .pre-commit-config.yaml:
```diff
@@ -7,35 +7,23 @@ repos:
     - repo: https://github.com/pre-commit/pre-commit-hooks
       rev: v4.5.0
       hooks:
-          - id: check-json
```
There are no json files in this repo

danielhollas commented on .pre-commit-config.yaml:
```diff
@@ -46,25 +34,3 @@ repos:
-          - id: isort
-            args: [--profile, black, --filter-files]
-
-    - repo: https://github.com/sirosen/check-jsonschema
```
This is checking the GHA yaml file against publicly available jsonschema. But the jsonschema is downloaded from the internet, so this check is slow.

yakutovicha replied:
Let's not remove the useful checks that are slow. I looked at the last commit - the maximum time I saw was two minutes.

While I am happy about the performance improvement, I would like it not to be at the cost of reducing the robustness.

danielhollas replied:
I am going to push back a little bit: as I understand it, this checks the correctness of the GHA yaml file against the specification. So it helps you catch issues when you change the file *before* you push to GitHub. 

But I don't think it reduces the robustness -- if you make a mistake, GitHub actions will yell at you anyway. I think by definition it cannot catch anything that GHA would not catch.

yakutovicha replied:
good point, agreed

danielhollas commented on .pre-commit-config.yaml:
```diff
@@ -46,25 +34,3 @@ repos:
                 - types-tabulate
                 - types-toml
-
-    - repo: https://github.com/asottile/setup-cfg-fmt
```
We'll probably move to `pyproject.toml` at some point so this is not as useful.

danielhollas commented on .pre-commit-config.yaml:
```diff
@@ -17,25 +17,13 @@ repos:
       hooks:
           - id: yamlfmt
 
-    - repo: https://github.com/mgedmin/check-manifest
```
This is apparently running `pip install` to build the package so it's quite slow. I think our MANIFEST file is stable enough now.

yakutovicha replied:
What about removing this in another PR together with switching to flit? IMO, flit does not require a MANIFEST file at all.

danielhollas replied:
Sure. I brought it back, but put it at the end because it's the slowest step. (I'll note that my worry is not so much about CI time, and more about local dev experience).

danielhollas commented on pyproject.toml:
```diff
@@ -0,0 +1,34 @@
+# TODO: Move entire setup.cfg to pyproject.toml
```
Ruff does not support reading configuration from `setup.cfg`. Instead of creating `ruff.toml`, I put the config into a minimal `pyproject.toml`, in anticipation of eventually moving setup.cfg to pyproject.toml. LMK if you disagree.

unkcpz replied:
Let's go, totally agree! Except for apps, it is safe to move other packages to `pyproject.toml` I think.

yakutovicha replied:
I think it is a good idea, yes.

danielhollas replied:
Okay, I'll open another PR for that, thanks!

unkcpz reviewed (approved):
Looks all good! Thanks @danielhollas

danielhollas commented:
Huh, interestingly, when I added back the `check-manifest` file, it started failing in pre-commit bot with

```python
['/pc/clone/LLopfktuRsCG39y1Br0Ylw/py_env-python3/bin/python', '-m', 'build', '--sdist', '.', '--outdir', '/tmp/check-manifest-64gqzu8s-sdist'] failed (status 1):
* Creating venv isolated environment...
ERROR Failed to create venv. Maybe try installing virtualenv.
  Command '/tmp/build-env-iai5zihp/bin/python3.11 -m ensurepip --upgrade --default-pip' failed with return code 1
  stdout:
    /tmp/build-env-iai5zihp/bin/python3.11: No module named ensurepip
```

It doesn't seem like an actual issue, just some missing dependencies. For now I've removed the build-related setup from `pyproject.toml` which seems to solve this. Let's deal with it in a follow-up PR.

yakutovicha reviewed (approved):
LGTM
